vpr/src/noc/sat_routing.cpp

```diff
@@ -0,0 +1,986 @@
+            }
+        }
+    }
+
+    return results;
+}
+
+static void forbid_illegal_turns(t_flow_link_var_map& flow_link_vars,
+                                 orsat::CpModelBuilder& cp_model) {
+    const auto& noc_ctx = g_vpr_ctx.noc();
+    const auto& traffic_flow_storage = noc_ctx.noc_traffic_flows_storage;
+
+    auto noc_routing_alg = dynamic_cast<const TurnModelRouting*> (noc_ctx.noc_flows_router.get());
+    // ensure that the routing algorithm is a turn model algorithm
+    VTR_ASSERT(noc_routing_alg != nullptr);
+
+    // forbid illegal turns based on the routing algorithm
+    // this includes 180 degree turns
+    for (const auto& [link1, link2] : noc_routing_alg->get_all_illegal_turns(noc_ctx.noc_model)) {
+        for (auto traffic_flow_id : traffic_flow_storage.get_all_traffic_flow_id()) {
+            auto& first_var = flow_link_vars[{traffic_flow_id, link1}];
+            auto& second_var = flow_link_vars[{traffic_flow_id, link2}];
+            // at most one of two consecutive links that form a turn can be activated
+            cp_model.AddBoolOr({first_var.Not(), second_var.Not()});
+        }
+    }
+}
+
+static vtr::vector<NocTrafficFlowId, int> quantize_traffic_flow_bandwidths(int bandwidth_resolution) {
+    const auto& noc_ctx = g_vpr_ctx.noc();
+    const auto& traffic_flow_storage = noc_ctx.noc_traffic_flows_storage;
+
+    //TODO: support heterogeneous bandwidth
+    const auto& noc_links = noc_ctx.noc_model.get_noc_links();
+    const double link_bandwidth = noc_links.front().get_bandwidth();
+    auto it = std::adjacent_find(noc_links.begin(), noc_links.end(), [](const NocLink& a, const NocLink& b){
+        return a.get_bandwidth() != b.get_bandwidth();
+    });
+
+    if (it != noc_links.end()) {
+        const NocLink& first_link = *it;
+        const NocLink& second_link = *(it + 1);
+        VTR_LOG_ERROR(
+            "SAT router assumes all NoC links have the same bandwidth. "
+            "NoC links %d and %d have different bandwidth: %g and %g",
+            (size_t)first_link.get_link_id(), (size_t)second_link.get_link_id(),
+            first_link.get_bandwidth(), second_link.get_bandwidth());
+    }
+
+    vtr::vector<NocTrafficFlowId, int> rescaled_traffic_flow_bandwidths;
+    rescaled_traffic_flow_bandwidths.resize(traffic_flow_storage.get_number_of_traffic_flows());
+
+    // rescale traffic flow bandwidths
+    for (auto traffic_flow_id : traffic_flow_storage.get_all_traffic_flow_id()) {
+        const auto& traffic_flow = traffic_flow_storage.get_single_noc_traffic_flow(traffic_flow_id);
+        double bandwidth = traffic_flow.traffic_flow_bandwidth;
+        int rescaled_bandwidth = (int)std::floor((bandwidth / link_bandwidth) * bandwidth_resolution);
+        rescaled_traffic_flow_bandwidths[traffic_flow_id] = rescaled_bandwidth;
+    }
+
+    return  rescaled_traffic_flow_bandwidths;
+}
+
+//static void add_congestion_constraints(t_flow_link_var_map& flow_link_vars,
```

Comment:
Group the placement functions together with a big banner above them saying they are only used for the SAT placer.

Reply:
Or delete if new code is better (I think you should it should be deleted).